awscli/topictags.py

```diff
@@ -0,0 +1,254 @@
+# SHALL THE AUTHOR BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+# WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+import os
+import json
+import docutils.core
+
+
+class TopicTagDB(object):
+    """This class acts like a database for the tags of all available topics.
+
+    A tag is an element in a topic reStructured text file that contains
+    information about a topic. Information can range from titles to even
+    related CLI commands. Here are all of the currently supported tags:
+
+    Tag                 Meaning                         Required?
+    ---                 -------                         ---------
+    :title:             The title of the topic          Yes
+    :description:       Sentence description of topic   Yes
+    :category:          Category topic falls under      Yes
+    :related topic:     A related topic                 No
+    :related command:   A related command               No
+
+    To see examples of how to specify tags looks in the directory
+    awscli/topics. Note that tags can have multiple values by delimiting
+    values with commas. All tags must be on their own line in the file.
+
+    This class can load a JSON index represeting all of topics and their tags,
+    scan all of the topics and store the values of their tags, retrieve the
+    tag value for a particular topic, query for all the topics with a specific
+    tag and/or value, and save the loaded data back out to a JSON index.
+
+    The structure of the database can be viewed as a python dictionary:
+
+    {'topic-name-1': {
+        'title': ['My First Topic Title'],
+        'description': ['This describes my first topic'],
+        'category': ['General Topics', 'S3'],
+        'related command': ['aws s3'],
+        'related topic': ['topic-name-2']
+     },
+     'topic-name-2': { .....
+    }
+
+    The keys of the dictionary are the CLI command names of the topics. These
+    names are based off the name of the reStructed text file that corresponds
+    to the topic. The value of these keys are dictionaries of tags, where the
+    tags are keys and their value is a list of values for that tag. Note
+    that all tag values for a specific tag of a specific topic are unique.
+    """
+
+    VALID_TAGS = ['category', 'description', 'title', 'related topic',
+                  'related command']
+
+    # The default directory to look for topics.
+    TOPIC_DIR = os.path.join(
+        os.path.dirname(
+            os.path.abspath(__file__)), 'topics')
+
+    # The default JSON index to load.
+    JSON_INDEX = os.path.join(TOPIC_DIR, 'topic-tags.json')
+
+    def __init__(self, tag_dictionary=None):
+        self._tag_dictionary = tag_dictionary
+        if self._tag_dictionary is None:
+            self._tag_dictionary = {}
+
+    def load_json_index(self, index_file=None):
```

Comment:
Why do we need that feature though?  You can just pass the index_file name into the load_json_index right?

Otherwise, do we need to support both interfaces?  Can we remove the `index_file` param?  Just looking to avoid adding things we don't actually need.